File: IPEDS-WD.py
# -*- coding: iso-8859-15 -*-
import csv
import json


def getQ(unitid):
    for i in crosswalk:
        if unitid == i[2]:
            return str(i[0])
    return "0"


with open('wikidata_query_result.json', 'rb') as f:
    data = json.loads(f.read())
    crosswalk = data["props"]["1771"]

countMatches = 0
with open('carnegieBasic.csv', 'rb') as csvfile:
    inflowFile = csv.DictReader(csvfile, delimiter=',', quotechar='"')
    for row in inflowFile:
        unitid = row['unitid']
        qid = getQ(unitid)
        if qid is not "0":
            targetData = row['Enrollment Profile']
            if targetData == "Exclusively graduate":
                targetQ = "Q23623496"
            elif targetData == "Exclusively undergraduate four-year":
                targetQ = "Q23623476"
            elif targetData == "Exclusively undergraduate two-year":
                targetQ = "Q23623474"
            elif targetData == "High undergraduate":
                targetQ = "Q23623483"
            elif targetData == "Majority graduate":
                targetQ = "Q23623491"
            elif targetData == "Majority undergraduate":
                targetQ = "Q23623487"
            elif targetData == "Very high undergraduate":
                targetQ = "Q23623479"                
            else:
                targetQ = None
            if targetQ is not None:
                countMatches += 1
                print("Q" + qid + "\t" + "P2643" + "\t" + targetQ +
                      "\t" + "S248" + "\t" + "Q4223026")

print("Identified {0} matches".format(countMatches))

File: IPEDS-WD.R
require(dplyr)

unitid.qid = read.csv('unitid-qid.csv')
unitid.qid$item = sub('http://www.wikidata.org/entity/','',unitid.qid$item)
colnames(unitid.qid) = c('qid','unitid')

carnegie.qid = read.csv('carnegie-wikidata-crosswalk.csv')

schools = read.csv('carnegieBasic.csv') %>%
  select(unitid, Undergraduate.Program, Graduate.Program,
         Enrollment.Profile, Undergraduate.Profile, Size...Setting, Basic)

crosswalked = merge(schools, unitid.qid, by.x = 'unitid', by.y = 'unitid')

#basic = merge(crosswalked, 
#              carnegie.qid %>% filter(carnegieClassificationType == 'Basic'),
#              by.x = 'Basic', by.y = 'carnegieClassification') %>%
#  select(schoolQid = qid, carnegieQid = wikidataQid)
#cat(paste(basic$schoolQid, 'P2643', basic$carnegieQid, 'S248','Q4223026',sep = '\t'), sep = '\n')

#enrollment.profile = merge(crosswalked, 
#                           carnegie.qid %>% filter(carnegieClassificationType == 'Enrollment Profile'),
#                           by.x = 'Enrollment.Profile', by.y = 'carnegieClassification') %>%
#  select(schoolQid = qid, carnegieQid = wikidataQid)
#cat(paste(enrollment.profile$schoolQid, 'P2643', enrollment.profile$carnegieQid, 'S248','Q4223026',sep = '\t'), sep = '\n')

#grad.program = merge(crosswalked, 
#                     carnegie.qid %>% filter(carnegieClassificationType == 'Graduate Instructional Program'),
#                     by.x = 'Graduate.Program', by.y = 'carnegieClassification') %>%
#  select(schoolQid = qid, carnegieQid = wikidataQid)
#cat(paste(grad.program$schoolQid, 'P2643', grad.program$carnegieQid, 'S248','Q4223026',sep = '\t'), sep = '\n')

#size.setting = merge(crosswalked, 
#                     carnegie.qid %>% filter(carnegieClassificationType == 'Size and Setting'),
#                     by.x = 'Size...Setting', by.y = 'carnegieClassification') %>%
#  select(schoolQid = qid, carnegieQid = wikidataQid)
#cat(paste(size.setting$schoolQid, 'P2643', size.setting$carnegieQid, 'S248','Q4223026',sep = '\t'), sep = '\n')

# undergraduate.profile = merge(crosswalked, 
#                               carnegie.qid %>% filter(carnegieClassificationType == 'Undergraduate Profile'),
#                               by.x = 'Undergraduate.Profile', by.y = 'carnegieClassification') %>%
#   select(schoolQid = qid, carnegieQid = wikidataQid)
# cat(paste(undergraduate.profile$schoolQid, 'P2643', undergraduate.profile$carnegieQid, 'S248','Q4223026',sep = '\t'), sep = '\n')

undergraduate.program = merge(crosswalked, 
                              carnegie.qid %>% filter(carnegieClassificationType == 'Undergraduate Program'),
                              by.x = 'Undergraduate.Program', by.y = 'carnegieClassification') %>%
  select(schoolQid = qid, carnegieQid = wikidataQid)
cat(paste(undergraduate.program$schoolQid, 'P2643', undergraduate.program$carnegieQid, 'S248','Q4223026',sep = '\t'), sep = '\n')